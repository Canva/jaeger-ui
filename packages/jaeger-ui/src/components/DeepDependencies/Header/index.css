/*
Copyright (c) 2019 Uber Technologies, Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

.DdgHeader {
  align-items: stretch;
  background: #fafafa;
  box-shadow: 0 2px 5px 0 rgba(0, 0, 0, 0.35);
  display: flex;
  flex-direction: column;
  flex: 1 1;
  position: relative;
  z-index: 10;
}

.DdgHeader--controlHeader {
  align-items: center;
  background-color: #f8f8f8;
  border-top: 1px solid #eee;
  display: flex;
  justify-content: space-between;
  padding-left: 1.25rem;
}

.DdgHeader--paramsHeader {
  flex-wrap: wrap;
  font-size: unset;
  align-items: center;
  background: #fff;
  display: flex;
  flex: 1;
  margin: 0;
  padding: 0.75rem 1.25rem 0.75rem 1.25rem;
  position: relative;
}

.DdgHeader--findWrapper {
  display: flex;
  flex: 1 1 auto;
  justify-content: flex-end;
}

.DdgHeader--uiFind {
  align-items: center;
  background-color: #fff;
  border-left: 1px solid #eee;
  border-right: 1px solid #eee;
  color: #bbb;
  display: flex;
  flex-grow: 1;
  max-width: 400px;
  min-width: 200px;
  transition: max-width 0.5s;
}

.DdgHeader--uiFind:focus-within {
  max-width: 550px;
}

.DdgHeader--uiFindSearchIcon {
  margin: 0 0.25em;
  padding: 0.24em;
}

.DdgHeader--uiFindInput > input:focus,
.DdgHeader--uiFindInput > input {
  border: none;
  box-shadow: none;
  border-radius: 0px;
}

.DdgHeader--uiFindCount {
  background: #ffffb8;
  border: 1px dashed #fadb14;
  border-radius: 4px;
  color: #ad8b00;
  margin-left: 0.25em;
  margin-right: 0.5em;
  transition: width 0.5s;
  padding: 0.27em 0.4em;
  stroke: #666;
  stroke-width: 1.2px;
  text-align: center;
}

.DdgHeader--uiFindCount:empty {
  display: none;
}
