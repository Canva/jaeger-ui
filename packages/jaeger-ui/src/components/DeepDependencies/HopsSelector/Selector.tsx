// Copyright (c) 2019 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import React, { PureComponent } from 'react';
import { Popover } from 'antd';
import cx from 'classnames';
import SortAmountAsc from 'react-icons/lib/fa/sort-amount-asc.js';
import IoChevronDown from 'react-icons/lib/io/chevron-down';
import IoChevronRight from 'react-icons/lib/io/chevron-right';

import { ECheckedStatus, EDirection, THop } from '../../../model/ddg/types';

import './Selector.css';

type TProps = {
  direction: EDirection;
  furthestDistance: number;
  furthestFullness: ECheckedStatus;
  handleClick: (distance: number, direction: EDirection) => void;
  hops: THop[];
};

const CLASSNAME = 'HopsSelector--Selector';

export default class Selector extends PureComponent<TProps> {
  private makeBtn = ({ distance, fullness, suffix }: THop & { suffix?: string }) => {
    const { handleClick, direction } = this.props;
    return (
      <button
        key={`${distance} ${direction} ${suffix}`}
        className={cx(`${CLASSNAME}--btn`, fullness, suffix || 'popover-content')}
        type="button"
        onClick={() => handleClick(distance, direction)}
      >
        {!suffix && <IoChevronRight className={`${CLASSNAME}--ChevronRight`} />} {Math.abs(distance)}
      </button>
    );
  };

  render() {
    const { direction, furthestDistance, furthestFullness, handleClick, hops } = this.props;
    const streamText = direction === EDirection.Downstream ? 'Down' : 'Up';
    const streamLabel = `${streamText}stream hops`;
    const lowercaseLabel = streamLabel.toLowerCase();
    if (hops.length === 1) {
      return <span className={CLASSNAME}>No {lowercaseLabel}</span>;
    }

    const decrementBtn = (
      <button
        key={`decrement ${direction}`}
        disabled={furthestDistance === 0}
        className={cx(CLASSNAME, 'decrement')}
        type="button"
        onClick={() => handleClick(furthestDistance - direction, direction)}
      >
        -
      </button>
    );
    const incrementBtn = (
      <button
        key={`increment ${direction}`}
        disabled={furthestDistance === hops[hops.length - 1].distance}
        className={cx(CLASSNAME, 'increment')}
        type="button"
        onClick={() => handleClick(furthestDistance + direction, direction)}
      >
        +
      </button>
    );
    const delimiterBtn = this.makeBtn({
      ...hops[hops.length - 1],
      suffix: 'delimiter',
    });
    const furthestBtn = this.makeBtn({
      distance: furthestDistance,
      fullness: furthestFullness,
      suffix: 'furthest',
    });

    return (
      <Popover
        arrowPointAtCenter
        content={[decrementBtn, ...hops.map(this.makeBtn), incrementBtn]}
        placement="bottom"
        visible={direction === EDirection.Upstream}
        title={`Visible ${lowercaseLabel}`}
      >
        <span className={CLASSNAME}>
          <SortAmountAsc className={cx(`${CLASSNAME}--AscIcon`, streamText)} /> {streamLabel} {furthestBtn} /{' '}
          {delimiterBtn} <IoChevronDown className={`${CLASSNAME}--ChevronDown`} />
        </span>
      </Popover>
    );
  }
}
