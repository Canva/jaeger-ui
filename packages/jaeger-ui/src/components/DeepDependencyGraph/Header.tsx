// Copyright (c) 2019 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import React, { PureComponent } from 'react';
import { Input } from 'antd';
import { History as RouterHistory, Location } from 'history';
import queryString from 'query-string';

// import './index.css';
import { TDdgActionMeta } from '../../model/ddg/types';

type TState = {
  end?: number;
  operation?: string;
  service?: string;
  start?: number;
};

type TProps = TState & {
  fetchDeepDependencyGraph: (query: TDdgActionMeta['query']) => void;
  history: RouterHistory;
  location: Location;
};

/*
 * This is a very crudely thrown-together component to act as a stand in until the header is properly
 * implemented. Currently this should suffice as a means to provide the arguments to fetch a ddg, but it is
 * far from an enjoyable experience to do so.
 */
export default class Header extends PureComponent<TProps, TState> {
  constructor(props: TProps) {
    super(props);
    this.state = {};

    const { service, operation, end = Date.now(), start: startProp } = this.props;
    const start = startProp || end - 60 * 60 * 1000;
    if (service != null && operation != null) {
      this.props.fetchDeepDependencyGraph({
        service,
        operation,
        end,
        start,
      });

      if (end !== this.props.end || start !== this.props.start) {
        const readOnlyQueryParams = queryString.parse(this.props.location.search);
        const queryParams = Object.assign({}, readOnlyQueryParams, { start, end });
        this.props.history.replace({
          ...this.props.location,
          search: `?${queryString.stringify(queryParams)}`,
        });
      }
    }
  }

  handleBlur = (name: keyof TState, evt: React.ChangeEvent<HTMLInputElement>) => {
    const { value } = evt.target;
    this.setState({ [name]: undefined });
    const readOnlyQueryParams = queryString.parse(this.props.location.search);
    const queryParams = Object.assign({}, readOnlyQueryParams, { [name]: value });
    const fetchArg: TState = {
      end: name === 'end' ? +value : this.props.end,
      start: name === 'start' ? +value : this.props.start,
      service: name === 'service' ? value : this.props.service,
      operation: name === 'operation' ? value : this.props.operation,
    };
    if (fetchArg.end == null) {
      fetchArg.end = Date.now();
      queryParams.end = `${fetchArg.end}`;
    }
    if (fetchArg.start == null) {
      fetchArg.start = fetchArg.end - 60 * 60 * 1000;
      queryParams.start = `${fetchArg.start}`;
    }
    // TODO: may need to check for both service and operation in early MVP
    if (fetchArg.service != null) {
      this.props.fetchDeepDependencyGraph(fetchArg as TDdgActionMeta['query']);
    }

    this.props.history.replace({
      ...this.props.location,
      search: `?${queryString.stringify(queryParams)}`,
    });
  };

  handleChange = (name: keyof TState, evt: React.ChangeEvent<HTMLInputElement>) => {
    const { value } = evt.target;
    this.setState({ [name]: value });
  };

  renderInput = (name: keyof TState, suffix?: string): React.ReactNode => {
    return (
      <form>
        <label htmlFor={name}>
          {name} {suffix}
        </label>
        <Input
          autosize={null}
          onBlur={this.handleBlur.bind(this, name) /* eslint-disable-line react/jsx-no-bind */}
          onChange={this.handleChange.bind(this, name) /* eslint-disable-line react/jsx-no-bind */}
          id={name}
          value={this.state[name] == null ? this.props[name] : this.state[name]}
        />
      </form>
    );
  };

  render() {
    return (
      <div>
        {this.renderInput('service')}
        {this.renderInput('operation')}
        {this.renderInput('start', 'in ms')}
        {this.renderInput('end', 'in ms')}
      </div>
    );
  }
}
