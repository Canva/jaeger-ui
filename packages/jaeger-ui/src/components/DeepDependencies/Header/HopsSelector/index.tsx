// Copyright (c) 2019 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import React, { PureComponent } from 'react';

import { decode, encodeDistance } from '../../../../model/ddg/visibility-codec';
import { ECheckedStatus, EDirection, TDdgModel, THop } from '../../../../model/ddg/types';
import Selector from './Selector';

type TProps = {
  ddgModel: TDdgModel;
  updateVisEncoding: (visEncoding: string) => void;
  visEncoding?: string;
};

export default class HopsSelector extends PureComponent<TProps> {
  private handleClick = (distance: number, direction: EDirection) => {
    const { ddgModel, updateVisEncoding, visEncoding } = this.props;

    updateVisEncoding(
      encodeDistance({
        ddgModel,
        direction,
        distance,
        prevVisEncoding: visEncoding,
      })
    );
  };

  render() {
    const { ddgModel, visEncoding } = this.props;
    const { distanceToPathElems } = ddgModel;

    const downstreamHops: THop[] = [];
    const upstreamHops: THop[] = [];
    let minVisDistance = 0;
    let minVisDistanceFullness = ECheckedStatus.Empty;
    let maxVisDistance = 0;
    let maxVisDistanceFullness = ECheckedStatus.Empty;
    const visibleIndices = visEncoding && new Set(decode(visEncoding));

    distanceToPathElems.forEach((elems, distance) => {
      let fullness: ECheckedStatus;
      if (visibleIndices) {
        const visible = elems.filter(({ visibilityIdx }) => visibleIndices.has(visibilityIdx));
        if (visible.length === elems.length) {
          fullness = ECheckedStatus.Full;
        } else if (!visible.length) {
          fullness = ECheckedStatus.Empty;
        } else {
          fullness = ECheckedStatus.Partial;
        }
      } else {
        fullness = Math.abs(distance) <= 2 ? ECheckedStatus.Full : ECheckedStatus.Empty;
      }

      if (distance >= 0) downstreamHops.push({ distance, fullness });
      if (distance <= 0) upstreamHops.push({ distance, fullness });

      if (fullness !== ECheckedStatus.Empty) {
        if (distance >= maxVisDistance) {
          maxVisDistance = distance;
          maxVisDistanceFullness = fullness;
        }
        if (distance <= minVisDistance) {
          minVisDistance = distance;
          minVisDistanceFullness = fullness;
        }
      }
    });

    downstreamHops.sort(({ distance: a }, { distance: b }) => a - b);
    upstreamHops.sort(({ distance: a }, { distance: b }) => b - a);

    return (
      <div>
        <Selector
          direction={EDirection.Upstream}
          handleClick={this.handleClick}
          hops={upstreamHops}
          furthestDistance={minVisDistance}
          furthestFullness={minVisDistanceFullness}
        />
        <Selector
          direction={EDirection.Downstream}
          handleClick={this.handleClick}
          hops={downstreamHops}
          furthestDistance={maxVisDistance}
          furthestFullness={maxVisDistanceFullness}
        />
      </div>
    );
  }
}
