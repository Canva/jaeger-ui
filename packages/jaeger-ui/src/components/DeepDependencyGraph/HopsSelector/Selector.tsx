// Copyright (c) 2019 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import React, { PureComponent } from 'react';
import { Popover } from 'antd';
import cx from 'classnames';

import { EStream, EFullness, THop } from './types';

import './Selector.css';

type TProps = {
  furthestDistance: number;
  furthestFullness: EFullness;
  handleClick: (distance: number, direction: EStream) => void;
  hops: THop[];
  stream: EStream;
};

export default class Selector extends PureComponent<TProps> {
  private btn = ({ distance, fullness, suffix = 'popover-content' }: THop & { suffix?: string }) => {
    const { handleClick, stream } = this.props;
    return (
      <button
        key={`${distance} ${stream} ${suffix}`}
        className={cx('Selector--btn', fullness, suffix)}
        type="button"
        onClick={() => handleClick(distance, stream)}
      >
        {Math.abs(distance)}
      </button>
    );
  };

  render() {
    const { furthestDistance, furthestFullness, handleClick, hops, stream } = this.props;
    const streamLabel = `${stream === EStream.down ? 'Down' : 'Up'}stream hops`;
    const lowercaseLabel = streamLabel.toLowerCase();
    if (hops.length === 1) {
      return <span className="Selector">No {lowercaseLabel}</span>;
    }

    const decrementBtn = (
      <button
        key={`decrement ${stream}`}
        disabled={furthestDistance === 0}
        type="button"
        onClick={() => handleClick(furthestDistance - stream, stream)}
      >
        -
      </button>
    );
    const incrementBtn = (
      <button
        key={`increment ${stream}`}
        disabled={furthestDistance === hops[hops.length - 1].distance}
        type="button"
        onClick={() => handleClick(furthestDistance + stream, stream)}
      >
        +
      </button>
    );
    const delimiterBtn = this.btn({
      ...hops[hops.length - 1],
      suffix: 'delimiter',
    });
    const furthestBtn = this.btn({
      distance: furthestDistance,
      fullness: furthestFullness,
      suffix: 'furthest',
    });

    return (
      <Popover
        arrowPointAtCenter
        content={[decrementBtn, hops.map(this.btn), incrementBtn]}
        placement="bottom"
        title={`Visible ${lowercaseLabel}`}
      >
        <span className="Selector">
          {streamLabel} {furthestBtn} / {delimiterBtn}
        </span>
      </Popover>
    );
  }
}
