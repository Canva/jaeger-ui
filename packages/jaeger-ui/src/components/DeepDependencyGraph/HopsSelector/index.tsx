// Copyright (c) 2019 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import React, { PureComponent } from 'react';

import { decode, encode } from '../../../model/ddg/visibility-codec';
import { TDdgModel } from '../../../model/ddg/types';
import Selector from './Selector';

import { EStream, EFullness, THop } from './types';

type TProps = {
  ddgModel: TDdgModel;
  updateVisEncoding: (visEncoding: string) => void;
  visEncoding?: string;
};

export default class HopsSelector extends PureComponent<TProps> {
  private handleClick = (clickedDistance: number, direction: EStream) => {
    const { ddgModel, updateVisEncoding, visEncoding } = this.props;
    const { distanceToPathElems, visIdxToPathElem } = ddgModel;

    let nextVisible: number[];
    if (visEncoding) {
      nextVisible = decode(visEncoding).filter(
        idx => visIdxToPathElem[idx] && Math.sign(visIdxToPathElem[idx].distance) !== direction
      );
    } else {
      nextVisible = [
        ...(distanceToPathElems.get(-1 * direction) || []),
        ...(distanceToPathElems.get(-2 * direction) || []),
      ].map(({ visibilityIdx }) => visibilityIdx);
    }

    for (let distance = 0; distance !== clickedDistance + direction; distance += direction) {
      const elems = distanceToPathElems.get(distance);
      if (elems) {
        nextVisible.push(...elems.map(({ visibilityIdx }) => visibilityIdx));
      }
    }

    updateVisEncoding(encode(nextVisible));
  };

  render() {
    const { ddgModel, visEncoding } = this.props;
    const { distanceToPathElems } = ddgModel;

    const hops: THop[] = [];
    let minVisDistance = 0;
    let minVisDistanceFullness = EFullness.empty;
    let maxVisDistance = 0;
    let maxVisDistanceFullness = EFullness.empty;
    const visibleIndices = visEncoding && new Set(decode(visEncoding));

    distanceToPathElems.forEach((elems, distance) => {
      let fullness: EFullness;
      if (visibleIndices) {
        const visible = elems.filter(({ visibilityIdx }) => visibleIndices.has(visibilityIdx));
        if (visible.length === elems.length) {
          fullness = EFullness.full;
        } else if (!visible.length) {
          fullness = EFullness.empty;
        } else {
          fullness = EFullness.partial;
        }
      } else {
        fullness = Math.abs(distance) <= 2 ? EFullness.full : EFullness.empty;
      }

      hops.push({ distance, fullness });
      if (fullness !== EFullness.empty) {
        if (distance > maxVisDistance) {
          maxVisDistance = distance;
          maxVisDistanceFullness = fullness;
        } else if (distance < minVisDistance) {
          minVisDistance = distance;
          minVisDistanceFullness = fullness;
        }
      }
    });

    hops.sort(({ distance: a }, { distance: b }) => a - b);
    const splitIdx = hops.findIndex(({ distance }) => !distance);
    const upstream = hops.slice(0, splitIdx + 1).reverse();
    const downstream = hops.slice(splitIdx);

    return (
      <div>
        <Selector
          handleClick={this.handleClick}
          hops={upstream}
          furthestDistance={minVisDistance}
          furthestFullness={minVisDistanceFullness}
          stream={EStream.up}
        />
        <Selector
          handleClick={this.handleClick}
          hops={downstream}
          furthestDistance={maxVisDistance}
          furthestFullness={maxVisDistanceFullness}
          stream={EStream.down}
        />
      </div>
    );
  }
}
