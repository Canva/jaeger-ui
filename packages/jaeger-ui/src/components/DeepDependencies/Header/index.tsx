// Copyright (c) 2019 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import * as React from 'react';

import HopsSelector from './HopsSelector';
import NameSelector from './NameSelector';
import { EDirection, TDdgDistanceToPathElems } from '../../../model/ddg/types';

import './index.css';

type TProps = {
  distanceToPathElems?: TDdgDistanceToPathElems;
  operation?: string;
  operationsForService: Record<string, string[]>;
  service?: string;
  services?: string[] | null;
  setDistance: (distance: number, direction: EDirection) => void;
  setOperation: (operation: string) => void;
  setService: (service: string) => void;
  visEncoding?: string;
};

export default class Header extends React.PureComponent<TProps> {
  render() {
    const {
      distanceToPathElems,
      operation,
      operationsForService,
      service,
      services,
      setDistance,
      setOperation,
      setService,
      visEncoding,
    } = this.props;

    return (
      <header className="DdgHeader">
        <div className="DdgHeader--paramsHeader">
          <NameSelector
            label="Service:"
            placeholder="Select a service…"
            value={service || null}
            setValue={setService}
            required
            options={services || []}
          />
          {service && (
            <NameSelector
              label="Operation:"
              placeholder="Select an operation…"
              value={operation || null}
              setValue={setOperation}
              required
              options={operationsForService[service] || []}
            />
          )}
        </div>
        <div className="DdgHeader--controlHeader">
          {distanceToPathElems && (
            <HopsSelector
              distanceToPathElems={distanceToPathElems}
              handleClick={setDistance}
              visEncoding={visEncoding}
            />
          )}
        </div>
      </header>
    );
  }
}
