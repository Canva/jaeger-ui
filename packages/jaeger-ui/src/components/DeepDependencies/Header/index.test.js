// Copyright (c) 2019 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import React from 'react';
import { shallow } from 'enzyme';

import Header from './index';
import HopsSelector from './HopsSelector';
import NameSelector from './NameSelector';

describe('<Header>', () => {
  const minProps = {
    setDistance: () => {},
    setOperation: () => {},
    setService: () => {},
  };
  const service = 'testService';
  const services = [service];
  const operation = 'testOperation';
  const operations = [operation];
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(<Header {...minProps} />);
  });

  it('renders with minimal props', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('omits the operation selector if a service is not selected', () => {
    const nameSelector = wrapper.find(NameSelector);
    expect(nameSelector.length).toBe(1);
    expect(nameSelector.prop('label')).toMatch(/service/i);
  });

  it('renders the operation selector if a service is selected', () => {
    let nameSelector = wrapper.find(NameSelector);
    wrapper.setProps({ service, services });
    nameSelector = wrapper.find(NameSelector);
    expect(nameSelector.length).toBe(2);
    expect(nameSelector.at(1).prop('label')).toMatch(/operation/i);
    expect(wrapper).toMatchSnapshot();

    wrapper.setProps({ operation, operations });
    expect(wrapper).toMatchSnapshot();
  });

  it('renders the hops selector if distanceToPathElems is provided', () => {
    wrapper.setProps({
      distanceToPathElems: new Map(),
      visEncoding: '3',
    });
    expect(wrapper.find(HopsSelector).length).toBe(1);
    expect(wrapper).toMatchSnapshot();
  });

  it('focuses uiFindInput IFF rendered when clicking on wrapping div', () => {
    const click = () => wrapper.find('.DdgHeader--uiFind').simulate('click');
    const focus = jest.fn();
    click();

    wrapper.instance()._uiFindInput = { current: { focus } };
    click();
    expect(focus).toHaveBeenCalledTimes(1);
  });
});
