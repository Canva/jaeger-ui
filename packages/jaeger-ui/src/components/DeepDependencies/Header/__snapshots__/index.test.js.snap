// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Header> renders the hops selector if distanceToPathElems is provided 1`] = `
<header
  className="DdgHeader"
>
  <div
    className="DdgHeader--paramsHeader"
  >
    <NameSelector
      label="Service:"
      options={Array []}
      placeholder="Select a service…"
      required={true}
      setValue={[Function]}
      value={null}
    />
  </div>
  <div
    className="DdgHeader--controlHeader"
  >
    <Component
      distanceToPathElems={Map {}}
      handleClick={[Function]}
      visEncoding="3"
    />
  </div>
</header>
`;

exports[`<Header> renders the operation selector if a service is selected 1`] = `
<header
  className="DdgHeader"
>
  <div
    className="DdgHeader--paramsHeader"
  >
    <NameSelector
      label="Service:"
      options={
        Array [
          "testService",
        ]
      }
      placeholder="Select a service…"
      required={true}
      setValue={[Function]}
      value="testService"
    />
    <NameSelector
      label="Operation:"
      options={Array []}
      placeholder="Select an operation…"
      required={true}
      setValue={[Function]}
      value={null}
    />
  </div>
  <div
    className="DdgHeader--controlHeader"
  />
</header>
`;

exports[`<Header> renders the operation selector if a service is selected 2`] = `
<header
  className="DdgHeader"
>
  <div
    className="DdgHeader--paramsHeader"
  >
    <NameSelector
      label="Service:"
      options={
        Array [
          "testService",
        ]
      }
      placeholder="Select a service…"
      required={true}
      setValue={[Function]}
      value="testService"
    />
    <NameSelector
      label="Operation:"
      options={Array []}
      placeholder="Select an operation…"
      required={true}
      setValue={[Function]}
      value="testOperation"
    />
  </div>
  <div
    className="DdgHeader--controlHeader"
  />
</header>
`;

exports[`<Header> renders with minimal props 1`] = `
<header
  className="DdgHeader"
>
  <div
    className="DdgHeader--paramsHeader"
  >
    <NameSelector
      label="Service:"
      options={Array []}
      placeholder="Select a service…"
      required={true}
      setValue={[Function]}
      value={null}
    />
  </div>
  <div
    className="DdgHeader--controlHeader"
  />
</header>
`;
